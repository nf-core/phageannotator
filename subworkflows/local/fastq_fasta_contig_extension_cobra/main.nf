//
// Extend contig ends using COBRA
//

include { BOWTIE2_BUILD                 } from '../../../modules/nf-core/bowtie2/build/main'
include { FASTQ_ALIGN_BOWTIE2           } from '../../nf-core/fastq_align_bowtie2/main'
include { GUNZIP                        } from '../../../modules/nf-core/gunzip/main'
include { COVERM_CONTIG as COVERM_COBRA } from '../../../modules/local/coverm/contig/main'
include { COBRAMETA                     } from '../../../modules/nf-core/cobrameta/main'

workflow FASTQ_FASTA_CONTIG_EXTENSION_COBRA {
    take:
    fastq_gz            // [ [ meta ], [ reads_1.fastq.gz, reads_2.fastq.gz ] ] , read files (mandatory)
    fasta_gz            // [ [ meta ], fasta.gz ]                               , assemblies/scaffolds (mandatory)
    viral_contigs_tsv   // [ [ meta ], virus_contigs.tsv ]                      , TSV file containing viral contig names
    cobra_assembler     // val:string {metaspades|megahit|idba}                 , string specifying which assembler was used
    mink                // val:int [ 0-999 ]                                    , integer specifying minimum kmer size used by assembler
    maxk                // val:int [ 0-999 ]                                    , integer specifying maximum kmer size used by assembler

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: Make bowtie2 index
    //
    ch_fasta_bt2 = BOWTIE2_BUILD ( fasta_gz ).index.first()
    ch_versions = ch_versions.mix( BOWTIE2_BUILD.out.versions )

    //
    // SUBWORKFLOW: Align reads to bowtie2 index
    //
    ch_fasta_alignment_bam = FASTQ_ALIGN_BOWTIE2 ( fastq_gz, ch_fasta_bt2, false, false, fasta_gz ).bam
    ch_versions = ch_versions.mix( FASTQ_ALIGN_BOWTIE2.out.versions )

    //
    // MODULE: Calculate abundance metrics from BAM file
    //
    ch_coverage_tsv = COVERM_COBRA ( ch_fasta_alignment_bam ).alignment_results
    ch_versions = ch_versions.mix( COVERM_COBRA.out.versions )

    // Remove header from coverage file
    ch_coverage_mod = ch_coverage_tsv
    .map { meta, tsv ->
        def tsv_mod = file("${workDir}/${meta.id}_coverage_mod.tsv")

        tsv.withReader { source ->
            tsv_mod.withWriter { target ->
                String line
                while( line=source.readLine() ) {
                    if ( !line.startsWith('Contig' )) {
                        target << line << '\n'
                    }
                }
            }
        }

        return [ meta, tsv_mod ]
    }

    //
    // MODULE: Gunzip FASTA file
    //
    ch_fasta = GUNZIP ( fasta_gz ).gunzip
    ch_versions = ch_versions.mix( GUNZIP.out.versions )

    // prepare input for cobra
    ch_cobra_input = ch_fasta
        .join(ch_coverage_mod)
        .join(viral_contigs_tsv)
        .join(ch_fasta_alignment_bam)
        .multiMap { it ->
            fasta: [ it[0], it[1] ]
            coverage: [ it[0], it[2] ]
            query: [ it[0], it[3] ]
            bam: [ it[0], it[4] ]
        }


    //
    // MODULE: Extend contigs using COBRA
    //
    ch_extended_fasta_gz = COBRAMETA (
        ch_cobra_input.fasta,
        ch_cobra_input.coverage,
        ch_cobra_input.query,
        ch_cobra_input.bam,
        cobra_assembler,
        mink,
        maxk
    ).all_cobra_assemblies
    ch_versions = ch_versions.mix( COBRAMETA.out.versions )

    emit:
    extended_fasta      = ch_extended_fasta_gz  // [ [ meta ], extended_contigs.fna.gz ]    , FASTA file containing extended contigs
    cobra_summary_tsv   = COBRAMETA.out.joining_summary
    versions            = ch_versions.unique()  // [ versions.yml ]
}
