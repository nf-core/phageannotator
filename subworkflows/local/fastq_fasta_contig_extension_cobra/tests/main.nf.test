nextflow_workflow {

    name "Test Subworkflow: FASTQ_FASTA_CONTIG_EXTENSION_COBRA"
    script "subworkflows/local/fastq_fasta_contig_extension_cobra/main.nf"
    workflow "FASTQ_FASTA_CONTIG_EXTENSION_COBRA"

    // Dependencies
    tag "BOWTIE2_BUILD"
    tag "FASTQ_ALIGN_BOWTIE2"
    tag "GUNZIP"
    tag "COVERM_CONTIG"
    tag "COBRAMETA"


    test("fastq - fasta - viruses_tsv - metaspades - k=22 - k=127") {

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz', checkifExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'test' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkifExists: true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id: 'test' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/tsv/test1_contigs_ids.tsv', checkifExists: true)
                    ]
                )
                input[3] = Channel.value( 'metaspades' )
                input[4] = Channel.value( 27 )
                input[5] = Channel.value( 127 )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.versions).match() },
                { assert path(workflow.out.extended_fasta[0][1]).linesGzip.any { it.contains(">k141_177_self_circular") } }
            )
        }
    }
}
