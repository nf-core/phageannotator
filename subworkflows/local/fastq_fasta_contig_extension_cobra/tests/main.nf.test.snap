{
    "fastq - fasta - viruses_tsv - metaspades - k=22 - k=127": {
        "content": [
            [
                "versions.yml:md5,2ae6371f345c3ed00f9a2abcdde4cd7c",
                "versions.yml:md5,3e74574fbfe0a90e2862245914d738e4",
                "versions.yml:md5,45ab424acfec849d28d8f8ab4a3af8f1",
                "versions.yml:md5,466c764b0ff156331507ba238e111bf2",
                "versions.yml:md5,64da2eecb8de1b06695a1d90486247b0",
                "versions.yml:md5,78a67cf0bf546994c775283ac774760a",
                "versions.yml:md5,7f61c2e84c088e886fef2f98710466b1",
                "versions.yml:md5,90c4e1c1c332a92141708a0be8ab2066",
                "versions.yml:md5,d464199cdf99a00939311c178e24c614",
                "versions.yml:md5,fc102771b3196a7b96fec4a6f1281a37"
            ]
        ],
        "timestamp": "2024-04-03T19:51:47.578170738"
    }
}